src/components/ActivityForm.js

```diff
@@ -27,57 +27,68 @@ export default class ActivityForm extends React.Component {
       name: props.activity ? props.activity.name : '',
       room: props.activity ? props.activity.room : '',
       teacher: props.activity ? props.activity.teacher : '',
-      error: '',
+      classNoError: '',
+      nameError: '',
+      roomError: '',
+      teacherError: '',
       // TODO: when you will use prop-types you will be able to improve this place ;)
       // TODO: I think, that now (looking at previous comments, maybe you will be able to
       // improve). If not, because activity will be able to be undefined, in separate PR
       // you should check lodash package and get method ;)
     };
   }
 
-  onNameChange = e => this.setState({ name: e.target.value });
+  onNameChange = e => {
+    this.setState({ name: e.target.value });
+    this.setState({ nameError: '' });
+  };
 
-  onTeacherChange = e => this.setState({ teacher: e.target.value });
+  onTeacherChange = e => {
+    this.setState({ teacher: e.target.value });
+    this.setState({ teacherError: '' });
+  };
 
-  onRoomChange = e => this.setState({ room: e.target.value });
+  onRoomChange = e => {
+    this.setState({ room: e.target.value });
+    this.setState({ roomError: '' });
+  };
 
   onClassNoChange = e => {
     const classNo = e.target.value;
     if (!classNo || classNo.match(/^[1-9]\d*$/)) {
       this.setState({ classNo });
     }
+    this.setState({ classNoError: '' });
   };
 
   onDayChange = e => this.setState({ day: e.target.value });
 
-  onBlur = e =>
-    !e.target.value && this.setState({ error: `Please provide ${e.target.placeholder}` });
+  onBlur = field => e =>
+    !e.target.value && this.setState({ [field]: `Please provide ${e.target.placeholder}` });
 
   onSubmit = e => {
     e.preventDefault();
+    this.setState({ nameError: '' });
+    this.setState({ classNoError: '' });
+    this.setState({ roomError: '' });
+    this.setState({ teacherError: '' });
     const { name, day, classNo, room, teacher } = this.state;
     if (!name || !day || !classNo || !teacher || !room) {
       // TODO: and then, this logic can be moved to Input.
       // You should have as many errors as required fields, separate ones.
-      let errorText = '';
       if (!name) {
-        errorText = `${errorText}Please provide Name. `;
-      }
-      if (!day) {
-        errorText = `${errorText}Please provide Day. `;
+        this.setState({ nameError: 'Please provide Name.' });
```

Comment:
at first errors should be collected to some object. If object has at least one error, then make a setState, now wil be:
- user clicks... 
- 4 rerenders to reset error.
- 4 rererenders if all errors exists.